File: LeetCode/Python3/Problem 139.py
# LeetCode: Problem 139
# Jonathan Kosasih

#Given a non-empty string s and a dictionary wordDict containing a list of non-empty words,
#determine if s can be segmented into a space-separated sequence of one or more dictionary words.
#
#Note:
#The same word in the dictionary may be reused multiple times in the segmentation.
#You may assume the dictionary does not contain duplicate words.
#
#Example 1:
#Input: s = "leetcode", wordDict = ["leet", "code"]
#Output: true
#Explanation: Return true because "leetcode" can be segmented as "leet code".
#
#Example 2:
#Input: s = "applepenapple", wordDict = ["apple", "pen"]
#Output: true
#Explanation: Return true because "applepenapple" can be segmented as "apple pen apple".
#             Note that you are allowed to reuse a dictionary word.
#             
#Example 3:
#Input: s = "catsandog", wordDict = ["cats", "dog", "sand", "and", "cat"]
#Output: false
#------------------------------------------------------------------
class Solution(object):
    def wordBreak(self, s, wordDict):
        """
        :type s: str
        :type wordDict: List[str]
        :rtype: bool
        """
        memo = [False] * len(s)
        for i in range(len(s)):
            for word in wordDict:
                # sliding window check for any words in word dictionary
                # mark last letter of window if it is a word AND the letter before it was true OR if this is the first word
                if(word == s[i-len(word)+1:i+1] and (memo[i-len(word)] or i-len(word) == -1)):
                    memo[i] = True
        # true only if the last letter makes a word, just report the last letter's status
        return memo[-1]

sol = Solution()
example = "leetcode"
wordDict = ["leet", "code"]
print(str(sol.wordBreak(example, wordDict)))

File: LeetCode/C++/Problem 139.cpp
// LeetCode: Problem 139
// Jonathan Kosasih
/*
Given a non-empty string s and a dictionary wordDict containing a list of non-empty words, determine if s can be segmented into a space-separated sequence of one or more dictionary words.

Note:

The same word in the dictionary may be reused multiple times in the segmentation.
You may assume the dictionary does not contain duplicate words.
*/
//-------------------------------------------------------------------

#include <iostream>
#include <vector>
#include <map>
#include <unordered_map>
#include <set>
#include <string>
#include <algorithm> 
#include <stack>
#include <queue>
#include <deque>
#include <numeric>
using std::vector;
using std::map;
using std::unordered_map;
using std::set;
using std::stack;
using std::queue;
using std::deque;
using std::accumulate;

class Solution {
public:
	bool wordBreak(string s, vector<string>& wordDict) {
		if (wordDict.size() == 0) {
			return false;
		}
		int stringLength = s.size();
		vector<bool> memo(stringLength + 1, false);
		memo[0] = true;

		for (int i = 0; i < stringLength; i++) {
			for (string& word : wordDict) {
				int wordLength = word.length();
				if (i+1-wordLength >= 0 && s.substr(i+1-wordLength,wordLength) == word && memo[i+1-wordLength]) {
					memo[i + 1] = true;
					break;
				}
			}
		}
		return memo.back();
	}
};
int main()
{
    //std::cout << "Hello World!\n";
}